### ISSUE TITLE
Next.js "PrivateRoute" issue

### ISSUE THREAD

**Issue by brunocrosier** [nhost/react-auth#3]
Following on from https://github.com/nhost/react-apollo/issues/4 

With the current setup, the `<NHostAuthProvider>` component requires Next.js users to wrap any pages where they are using `useQuery`, `useMutation` or `useSubscription` with a `PrivateRoute` HOC (e.g https://github.com/nhost-examples/nextjs-apollo/blob/main/components/private-route.jsx)

Currently this component is only available in the Nhost examples. One option could be to export the `PrivateRoute` component directly from `@nhost/react-auth`

However, I think this still leads to a not great developer experience, because it requires wrapping the pages in the HOC every time

@elitan raised the point on the other issue:

> I think the reason why [official Next.js + Apollo examples](https://github.com/vercel/next.js/tree/canary/examples/with-apollo) does not use a similar `PrivateRoute` approach is because it does not handle auth.

I am not 100% sure, but I think it's definitely possible to handle auth + Apollo with Next.js, without the wrapper approach (see [this official example](https://github.com/vercel/next.js/blob/canary/examples/api-routes-apollo-server-and-client-auth/))

If we want to keep the HOC strategy, I also agree that the name `PrivateRoute` could be improved. Maybe `AuthProtected` or `WithAuthProtection` 🤷‍♂️

**Comment by elitan** [comment]
> However, I think this still leads to a not great developer experience, because it requires wrapping the pages in the HOC every time

Is there any alternative solution for Next.js?

> I am not 100% sure, but I think it's definitely possible to handle auth + Apollo with Next.js, without the wrapper approach (see this official example)

They seem to use a cookie approach to authenticate graphql request. Hasura requires a JWT token to be sent instead. I don't see anywhere client side where they check if the user is logged in or not like we do in the `PrivateRoute` component.

Another suggestion: `withProtectedAuth` from @leerob.

**Comment by brunocrosier** [comment]
Hmm I can't find anything about `withProtectedAuth` online, could you send a link?

>They seem to use a cookie approach to authenticate graphql request. Hasura requires a JWT token to be sent instead. I don't see anywhere client side where they check if the user is logged in or not like we do in the `PrivateRoute` component.

Ah, what about looking to [next-auth](https://next-auth.js.org/configuration/options#jwt)

There is also [this Next.js + Hasura boilerplate app](https://github.com/ghoshnirmalya/nextjs-hasura-boilerplate) which seems to be handling everything without the HOC

**Comment by leerob** [comment]
There’s nothing online, I just said that seems like a decent name. It’s common with prefix HOCs with “with” 👍🏼

**Comment by brunocrosier** [comment]
> There’s nothing online, I just said that seems like a decent name. It’s common to prefix HOCs with “with” 👍🏼

Oh, my bad! 😅

**Comment by leerob** [comment]
No worries 😄

**Comment by brunocrosier** [comment]
Just to update this a little..

The main problem with the `PrivateRoute` solution is that it means that Next.js developers using Nhost can no longer benefit from Server-Side Rendering (the initial SSR HTML source will always be "checking auth...")

Ideally, I think we should allow developers to continue leveraging SSR methods like `getServerSideProps` in Next.js to fetch data on the server side.

A really good example of how this can be achieved is available [here](https://github.com/nextauthjs/next-auth-example/blob/308a1f127eb9f1ad079765b407d210b5b17e3fd9/pages/server.js#L31). 

I managed to hack together something locally which uses [next-auth](next-auth.js.org) and the ideas from [Next.js + Hasura boilerplate](https://github.com/ghoshnirmalya/nextjs-hasura-boilerplate), and it more or less works with Nhost.

The problem is that this solution requires editing the database schema for the "users" table within Hasura (as well as adding tables for verification requests etc. This is necessary because of how `next-auth` works. However, this breaks the Nhost "Auth" section (see screenshot below).

![image](https://user-images.githubusercontent.com/18399089/106276609-c00d1580-622f-11eb-9bed-6fa4171dec8c.png)

What are your thoughts about this? I would personally recommend trying to find a way to make `next-auth` play nice with Nhost. It's also worth noting that Hasura did a livestream 2 days ago in which they started building a Next.js + Hasura fullstack app, and you can see [here](https://youtu.be/CICBihCe7hQ?t=4833) that they mention that in the second installment of the series (next week), they will add auth with `next-auth`.

**Comment by elitan** [comment]
My recommendation is to skip SSR when you need authentication and only check it client side.

It has been discussed before here: https://github.com/nhost/hasura-backend-plus/issues/280

**Comment by brunocrosier** [comment]
Ultimately this is up to you, but I think it would be a shame to miss out on this feature when it is technically possible. Also, if Hasura will start recommending `next-auth` for Hasura + Next.js as soon as next week, it could lead to a lot of confusion in future when Nhost users come across this obstacle.

In the meantime, I would recommend updating the Nhost documentation to explain that a custom `PrivateRoute` component is needed in this scenario

**Comment by elitan** [comment]
Yea my recommendation is up to me. Just to clarify, I'm not saying you (or anyone else) must follow them nor does it stop anyone from creating a auth implementation for SSR with Next.js.

From our side it's a mix of not seeing the need from our selfs nor our customers together with limited time on our end. If anyone (you?) want to dig deeper and solve this it would be amazing.

> In the meantime, I would recommend updating the Nhost documentation to explain that a custom PrivateRoute component is needed in this scenario

Agreed.

**Comment by brunocrosier** [comment]
For sure. I am very happy to contribute and to try and find a better solution. I've spent a good few hours hacking together this proof-of-concept that uses next-auth.

Essentially, I suspect that in order for this to all work smoothly in future, there would need to be a step in HBP / nhost when creating a project, which allows the developer to select between create-react-app and next.js

If next.js is selected, HBP would need to add some extra database columns in order to accommodate next-auth.

Basically, I would be happy to dedicate some time to it if you think that it's a good idea

**Comment by elitan** [comment]
Ok. So I'll close this issue for now since the component `PrivateRoute` works.

If you want to add features please add a PR or an issue with specifications 👍

**Comment by noverby** [comment]
Just for reference. Here is my solution to avoid HOC:
```typescript
import React from "react";
import { useAuth } from "@nhost/react-auth";

export default function Layout({ children }: { children?: any }) {
  const { signedIn } = useAuth();

  return (
    <>
      {signedIn != null && children}
    </>
  );
}
```
